presto-main/src/main/java/io/prestosql/operator/scalar/DateTimeFunctions.java

```diff
@@ -139,6 +146,28 @@ public static long fromUnixTime(@SqlType(StandardTypes.DOUBLE) double unixTime,
         return packDateTimeWithZone(Math.round(unixTime * 1000), zoneId.toStringUtf8());
     }
 
+    @ScalarFunction("from_unixtime_nanos")
+    @SqlType("timestamp(9)")
+    public static LongTimestamp fromUnixTimeNanos(@SqlType(StandardTypes.BIGINT) long unixTimeNanos)
+    {
+        long epochSeconds = unixTimeNanos / Timestamps.NANOSECONDS_PER_SECOND;
+        long nanosOfSecond = unixTimeNanos % Timestamps.NANOSECONDS_PER_SECOND;
+        long picosOfSecond = Timestamps.PICOSECONDS_PER_NANOSECOND * nanosOfSecond;
+        return DateTimes.longTimestamp(epochSeconds, picosOfSecond);
+    }
+
+    @ScalarFunction("from_unixtime_nanos")
+    @LiteralParameters({"p", "s"})
+    @SqlType("timestamp(9)")
+    public static LongTimestamp fromUnixTimeNanos(@LiteralParameter("s") long numScale, @SqlType("decimal(p, s)") Slice unixTimeNanosDecimal)
```

Comment:
You also need a version takes short version of decimal, which is represented as `long`.
Otherwise call like this one `from_unixtime_nanos(DECIMAL '123456789)` will fail.

You can group two implementation in internal static class. See https://github.com/prestosql/presto/blob/2379c52ead0ca9f50d89fd75aa25e885a81118fa/presto-main/src/main/java/io/prestosql/operator/scalar/MathFunctions.java#L140-L166 as an example

Reply:
is there somewhere where the implementation details of the short version of decimal are written up? specifically i'm wondering how a non-integer SHORT decimal is represented (e.g. `DECIMAL '123.456'`)

Reply:
The current branch works for integer SHORT decimals but since we want to now support either rounding or throwing on non-integer DECIMALS I'll now need to worry about this.

Reply:
> i'm wondering how a non-integer SHORT decimal is represented

I am not sure if it is written up anywhere. We basically keep the digits and the sign in the value. And where the decimal dot should be put is derived from type parameters.